Push rule tweaks tests

Fizzadar opened the pull request:
I've been digging into some mysterious push notifications we're not expecting to see and ended up adding this test & tweaking the variable (haven't got to the bottom of our issue!). Figured both may be good additions so PR-ing, can split into two if desired.

Signed off by Nick Mills-Barrett <nick@beeper.com>

### Pull Request Checklist

<!-- Please read https://matrix-org.github.io/synapse/latest/development/contributing_guide.html before submitting your pull request -->

* [x] Pull request is based on the develop branch
* [x] Pull request includes a [changelog file](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#changelog).
* [x] Pull request includes a [sign off](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#sign-off)
* [x] [Code style](https://matrix-org.github.io/synapse/latest/code_style.html) is correct
  (run the [linters](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#run-the-linters))

clokep commented on synapse/push/bulk_push_rule_evaluator.py:
```diff
@@ -274,7 +274,7 @@ def _condition_checker(
 ) -> bool:
     for cond in conditions:
-        _id = cond.get("_id", None)
+        _id = cond.get("_cache_key", None)
         if _id:
             res = cache.get(_id, None)
```
Probably worth making the `_id` variable into `_cache_key` also?

Fizzadar replied:
https://github.com/matrix-org/synapse/pull/12188/commits/3e8dac0dcc25ddedcfff7c6dcdb457599d038f57

clokep commented on tests/push/test_push_rules_http.py:
```diff
@@ -0,0 +1,117 @@
+
+        return hs
+
+    def _make_user_with_pusher(self, username):
```
We've been trying to slow get type hints everywhere, especially in easy spots:

```suggestion
    def _make_user_with_pusher(self, username: str) -> Tuple[str, str]:
    	"""
    	Register a user, configure an HTTP pusher for them.
    	
    	Return:
    		A tuple of the user name & access token.
    	"""
```

Fizzadar replied:
https://github.com/matrix-org/synapse/pull/12188/commits/e10edf91bfc1da7b29d6cdc90d874d8fb24ab96d

clokep commented on tests/push/test_push_rules_http.py:
```diff
@@ -0,0 +1,117 @@
+        # The other user joins
+        self.helper.join(room=room, user=other_user_id, tok=other_access_token)
+
+        # The other user sends a message (ignored by dont_notify push rule set above)
+        self.helper.send(room, body="Hi!", tok=other_access_token)
+
+        # The user sends a message back (sends a notification)
+        self.helper.send(room, body="Hello", tok=access_token)
+
+        self.assertEqual(len(self.push_attempts), 1)
```
We should probably assert that it is 0 after the first message is sent, so we know which of the above messages is causing the push:

```suggestion
        # The other user sends a message (ignored by dont_notify push rule set above)
        self.helper.send(room, body="Hi!", tok=other_access_token)
        self.assertEqual(len(self.push_attempts), 0)

        # The user sends a message back (sends a notification)
        self.helper.send(room, body="Hello", tok=access_token)
        self.assertEqual(len(self.push_attempts), 1)
```

(We even might want to do it before sending the first event too?)

Fizzadar replied:
https://github.com/matrix-org/synapse/pull/12188/commits/39b8758777a6cde5d34b74072a36a7deaa59ea73

clokep commented on tests/push/test_push_rules_http.py:
```diff
@@ -0,0 +1,117 @@
+# Copyright 2018 New Vector
```
```suggestion
# Copyright 2022 The Matrix.org Foundation C.I.C.
```

(Or your own name or Beeper.)

clokep commented on tests/push/test_push_rules_http.py:
```diff
@@ -0,0 +1,117 @@
+from tests.unittest import HomeserverTestCase
+
+
+class HTTPPusherTests(HomeserverTestCase):
```
Not a requirement, but if this and `HTTPPusherTests` share a lot of code (or could) it can be nice to have an abstract base class they both inherit from.

It really isn't 100% clear to me why this is a different class anyway? Should this be renamed or maybe have comments added about who they differ?

Fizzadar replied:
Totally agree - was originally separate for ease of writing / testing it, I've merged it back into the existing module in https://github.com/matrix-org/synapse/pull/12188/commits/26121c719c873ff74e912c69ac8edf59380a1e5b.

clokep reviewed (commented):
I left a few comments, I *think* this change would be OK, as the `_id` property is fully internal, but not 100% sure on that!

Fizzadar commented:
Thanks for the feedback @clokep! I've addressed the comments. Looking across the codebase I can't see any use of the `_cache_key`/`_id` filed anywhere so believe this is a safe change!

clokep reviewed (approved):
LGTM! Thank you!

If in your journey of debugging you find it helpful to add any comments / docstrings to these methods, it would certainly help with understanding how it all fits together!
